pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -142,80 +141,77 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, pr *v1beta1.PipelineRun)
-			logger.Errorf("Failed to create tracker for TaskRuns for PipelineRun %s: %v", pr.Name, err)
-			recorder.Event(pr, corev1.EventTypeWarning, v1beta1.PipelineRunReasonFailed.String(), "Failed to create tracker for TaskRuns for PipelineRun")
-			return err
-		}
+		err := cancelPipelineRun(logger, pr, c.PipelineClientSet)
+		return c.finishReconcileUpdateEmitEvents(ctx, pr, before, err)
+	}
 
-		// Make sure that the PipelineRun status is in sync with the actual TaskRuns
-		err := c.updatePipelineRunStatusFromInformer(ctx, pr)
-		if err != nil {
-			// This should not fail. Return the error so we can re-try later.
-			logger.Errorf("Error while syncing the pipelinerun status: %v", err.Error())
-			return err
-		}
+	if err := c.tracker.Track(pr.GetTaskRunRef(), pr); err != nil {
+		logger.Errorf("Failed to create tracker for TaskRuns for PipelineRun %s: %v", pr.Name, err)
+		return c.finishReconcileUpdateEmitEvents(ctx, pr, before, err)
+	}
 
-		// Reconcile this copy of the pipelinerun and then write back any status or label
-		// updates regardless of whether the reconciliation errored out.
-		if err = c.reconcile(ctx, pr); err != nil {
-			logger.Errorf("Reconcile error: %v", err.Error())
-			merr = multierror.Append(merr, err)
-		}
+	// Make sure that the PipelineRun status is in sync with the actual TaskRuns
+	err := c.updatePipelineRunStatusFromInformer(ctx, pr)
+	if err != nil {
+		// This should not fail. Return the error so we can re-try later.
+		logger.Errorf("Error while syncing the pipelinerun status: %v", err.Error())
+		return c.finishReconcileUpdateEmitEvents(ctx, pr, before, err)
 	}
 
-	// If we need to update the labels or annotations, we need to call Update with these
-	// changes explicitly.
-	if !reflect.DeepEqual(original.ObjectMeta.Labels, pr.ObjectMeta.Labels) || !reflect.DeepEqual(original.ObjectMeta.Annotations, pr.ObjectMeta.Annotations) {
-		if _, err := c.updateLabelsAndAnnotations(pr); err != nil {
-			logger.Warn("Failed to update PipelineRun labels/annotations", zap.Error(err))
-			recorder.Event(pr, corev1.EventTypeWarning, "Error", "PipelineRun failed to update labels/annotations")
-			return multierror.Append(merr, err)
-		}
+	// Reconcile this copy of the pipelinerun and then write back any status or label
+	// updates regardless of whether the reconciliation errored out.
+	if err = c.reconcile(ctx, pr); err != nil {
+		logger.Errorf("Reconcile error: %v", err.Error())
+	}
+
+	return c.finishReconcileUpdateEmitEvents(ctx, pr, before, err)
+}
+
+func (c *Reconciler) finishReconcileUpdateEmitEvents(ctx context.Context, pr *v1beta1.PipelineRun, beforeCondition *apis.Condition, previousError error) error {
+	recorder := controller.GetEventRecorder(ctx)
+	logger := logging.FromContext(ctx)
+
+	afterCondition := pr.Status.GetCondition(apis.ConditionSucceeded)
+	events.Emit(recorder, beforeCondition, afterCondition, pr)
+	_, err := c.updateLabelsAndAnnotations(pr)
+	if err != nil {
```

Comment:
no issues here with these changes but this same check is missing from the taskrun `finishReconcileUpdateEmitEvents`:

```
	_, err := c.updateLabelsAndAnnotations(tr)
	events.EmitError(recorder, err, tr)
```

`taskrun` emits event without checking if err is `nil`, we can create a separate PR for that

Reply:
Just read the [comments](https://github.com/tektoncd/pipeline/pull/2526#discussion_r424027814) from the taskrun PR, if `EmitError` has check for `nil`, should we drop the check from here like taskrun reconciler? Or do we still want to keep it?

Reply:
if we drop `nil` check from here, no need for the `logger` as well ...

Reply:
I added the check here because @bobcatfish mentioned before that she didn't like the check to be in `EmitError` only, so I was planning on changing that in the task run controller too (but in a different PR)

Reply:
thanks @afrittoli